File: app/graph/state.py
"""State schemas for LangGraph orchestrator."""

from datetime import datetime, timezone
from enum import StrEnum
from typing import TypedDict


class Session(StrEnum):
    CLOSED = "CLOSED"
    PRE_MARKET = "PRE_MARKET"
    SESSION_1 = "SESSION_1"
    LUNCH = "LUNCH"
    SESSION_2 = "SESSION_2"
    AFTER_HOURS = "AFTER_HOURS"


class SignalType(StrEnum):
    PRICE_MOVE = "PRICE_MOVE"
    VOLUME_SPIKE = "VOLUME_SPIKE"
    COMBINED = "COMBINED"


class SignalTier(StrEnum):
    MINOR = "MINOR"
    MAJOR = "MAJOR"


class Depth(StrEnum):
    LIGHT = "LIGHT"
    FULL = "FULL"
    DEEP = "DEEP"


class TickerSignal(TypedDict):
    ticker: str
    signal_type: SignalType
    tier: SignalTier
    value: float
    detected_at: datetime


class TickerResult(TypedDict):
    snapshot_id: int
    recommendation: str
    clean_html: str
    signals: list[TickerSignal]
    sent: bool


class OrchestratorState(TypedDict):
    current_session: Session
    last_check: datetime
    signals: list[TickerSignal]
    pending_batch: list[str]
    last_run: dict[str, datetime]
    last_news_fetch: datetime
    _route: str


class AnalysisState(TypedDict):
    tickers: list[str]
    depth: Depth
    session: Session
    results: dict[str, TickerResult]
    errors: dict[str, str]


def initial_orchestrator_state() -> OrchestratorState:
    return OrchestratorState(
        current_session=Session.CLOSED,
        last_check=datetime.now(timezone.utc),
        signals=[],
        pending_batch=[],
        last_run={},
        last_news_fetch=datetime.min.replace(tzinfo=timezone.utc),
        _route="skip",
    )

File: app/src/graph/state.ts
// app/src/graph/state.ts
export type Session = 'CLOSED' | 'PRE_MARKET' | 'SESSION_1' | 'LUNCH' | 'SESSION_2' | 'AFTER_HOURS'
export type SignalType = 'PRICE_MOVE' | 'VOLUME_SPIKE' | 'COMBINED'
export type SignalTier = 'MINOR' | 'MAJOR'
export type Depth = 'LIGHT' | 'FULL' | 'DEEP'

export interface TickerSignal {
  ticker: string
  signal_type: SignalType
  tier: SignalTier
  value: number
  detected_at: string
}

export interface OrchestratorState {
  current_session: Session
  last_session: Session | null
  last_check: string
  last_scheduled: string | null
  signals: TickerSignal[]
  signal_cooldowns: Record<string, string>
  pending_batch: string[]
  last_run: string | null
  last_news_fetch: string | null
  _route?: string
}

export interface AnalysisState {
  tickers: string[]
  depth: Depth
  session: Session
  alerts: 'spike' | 'dedup' | 'silent'
  results: Record<string, string>
  errors: Record<string, string>
}
